File: Tests/SourceStateTests.cs
using System;
using System.Text.Patterns;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Tests {
	[TestClass]
	public class SourceStateTests {
		[TestMethod]
		public void StoreAndRestoreTest() {
			Source source = new Source("hello world");
			SourceState state = source.Store();
			ResultAssert.Succeeds("hello".Consume(ref source));
			source.Restore(ref state);
			//The only way this can succeed is if the restore worked
			ResultAssert.Succeeds("hello".Consume(ref source));
		}
	}
}

File: Patterns.Tests/SourceStateTests.fs
namespace Patterns

open System
open Stringier.Patterns
open Microsoft.VisualStudio.TestTools.UnitTesting

[<TestClass>]
type SourceStateTests() =

    [<TestMethod>]
    member _.``store and restore`` () =
        let mutable source = Source("hello world")
        let state = source.Store()
        ResultAssert.Succeeds("hello".Consume(&source))
        source.Restore(state)
        //The only way this can succeed is if the restore worked
        ResultAssert.Succeeds("hello".Consume(&source))